source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc

```diff
@@ -226,6 +228,16 @@ class GrpcStatsFilter : public Http::PassThroughFilter {
     if (doStatTracking()) {
       config_->context_.chargeStat(*cluster_, Grpc::Context::Protocol::Grpc, request_names_,
                                    trailers.GrpcStatus());
+
+      if (config_->enable_upstream_stats_ &&
+          decoder_callbacks_->streamInfo().lastUpstreamTxByteSent().has_value() &&
+          decoder_callbacks_->streamInfo().lastUpstreamRxByteReceived().has_value()) {
+        std::chrono::milliseconds chrono_duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                decoder_callbacks_->streamInfo().lastUpstreamRxByteReceived().value() -
+                decoder_callbacks_->streamInfo().lastUpstreamTxByteSent().value());
+        config_->context_.chargeUpstreamStat(*cluster_, request_names_, chrono_duration);
+      }
```

Comment:
@ggreenway @Sh4d1 post review drive by: this would not include "trailers only" responses (headers only responses with an error). I think this is probably not what we want and we should include error responses also? Can we fix/clarify this in a follow up? Thank you.

Reply:
@mattklein123 good catch! I think adding the same logic in encodeHeaders if `end_stream` is true should be enough right? I'll send the fix tomorrow I think 😄

Reply:
Or move the whole logic there 🤔